test/integration/endpoints/endpoints_test.go

```diff
@@ -156,6 +156,161 @@ func TestEndpointUpdates(t *testing.T) {
+		informers.Core().V1().Pods(),
+		informers.Core().V1().Services(),
+		informers.Core().V1().Endpoints(),
+		client,
+		0)
+
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go epController.Run(1, stopCh)
+
+	// Create namespace
+	ns := framework.CreateTestingNamespace("test-endpoints-terminating", server, t)
+	defer framework.DeleteTestingNamespace(ns, server, t)
+
+	// Create a pod with labels
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "test-pod",
+			Labels: labelMap(),
+		},
+		Spec: v1.PodSpec{
+			NodeName: "fake-node",
+			Containers: []v1.Container{
+				{
+					Name:  "fakename",
+					Image: "fakeimage",
+					Ports: []v1.ContainerPort{
+						{
+							Name:          "port-443",
+							ContainerPort: 443,
+						},
+					},
+				},
+			},
+		},
+		Status: v1.PodStatus{
+			Phase: v1.PodRunning,
+			Conditions: []v1.PodCondition{
+				{
+					Type:   v1.PodReady,
+					Status: v1.ConditionTrue,
+				},
+			},
+			PodIP: "10.0.0.1",
+			PodIPs: []v1.PodIP{
+				{
+					IP: "10.0.0.1",
+				},
+			},
+		},
+	}
+
+	createdPod, err := client.CoreV1().Pods(ns.Name).Create(context.TODO(), pod, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create pod %s: %v", pod.Name, err)
+	}
+
+	createdPod.Status = pod.Status
+	_, err = client.CoreV1().Pods(ns.Name).UpdateStatus(context.TODO(), createdPod, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to update status of pod %s: %v", pod.Name, err)
+	}
+
+	// Create a service associated to the pod
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-service",
+			Namespace: ns.Name,
+			Labels: map[string]string{
+				"foo": "bar",
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{
+				"foo": "bar",
+			},
+			Ports: []v1.ServicePort{
+				{Name: "port-443", Port: 443, Protocol: "TCP", TargetPort: intstr.FromInt(443)},
+			},
+		},
+	}
+	_, err = client.CoreV1().Services(ns.Name).Create(context.TODO(), svc, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create service %s: %v", svc.Name, err)
+	}
+
+	// poll until associated Endpoints to the previously created Service exists
+	if err := wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+		endpoints, err := client.CoreV1().Endpoints(ns.Name).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		numEndpoints := 0
+		for _, subset := range endpoints.Subsets {
+			numEndpoints += len(subset.Addresses)
+		}
+
+		if numEndpoints == 0 {
+			return false, nil
+		}
+
+		return true, nil
+	}); err != nil {
+		t.Fatalf("endpoints not found: %v", err)
+	}
+
+	err = client.CoreV1().Pods(ns.Name).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("error deleting test pod: %v", err)
+	}
+
+	// poll until endpoint for deleted Pod is no longer in Endpoints.
```

Comment:
How do you differentiate between deleted and terminated here?

Should you keep the pod in terminating state for a long time and check the address is not there during that time?

Reply:
I had the same question when writing this test. And in this particular integration test, a deleted pod is never actually removed from the system, I think it's because we're not running the garbage collector controller but I'm not 100% sure about that. Either way the pod will exist with a deletion timestamp for the duration of this test at least.

Reply:
This is where a E2E test might add more value, because the termination grace period we set on the pod will define this behaviour.

Reply:
What happens if we add a finalizer to the pod? It won't be deleted but it won't be ready, right?
I thnk there are others tests in integration that simulates this ... I'll take a look next week

Reply:
> What happens if we add a finalizer to the pod? It won't be deleted but it won't be ready, right?

In this integration test specifically, the pod is never actually deleted. But trying to delete the pod sets the deletion timestamp which is what endpoints controller is watching to exclude terminating endpoints.

Reply:
Is not what we want?
To ensure terminating endpoints are not present?

Reply:
In this test we want pods to exist with a deletion timestamp, but its associated endpoint should not exist. And this is the current behavior of this test right now.

Reply:
ah, ok, so I thought the `client.CoreV1().Pods(ns.Name).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})` will delete the pod inmeditialy, but I see it is only setting the deletion timestamp.
That was the part I was missing, sorry

Reply:
fwiw I think it wouldn't hurt to assert the deletion timestamp behavior here so we know we're not testing against a fully deleted pod. Will try to circle back with that

Reply:
heh, that is what I did to test it :smile:

Reply:
Added an extra assertion here https://github.com/kubernetes/kubernetes/pull/103645